flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterClientTest.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.flink.runtime.minicluster.MiniCluster;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.concurrent.CompletableFuture;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link PerJobMiniClusterClient}.
+ */
+public class PerJobMiniClusterClientTest {
+
+	@Test
+	public void testClusterShutdown() {
+		runClusterShutdownTest(false);
+	}
+
+	@Test
+	public void testClusterShutdownWithException() {
+		runClusterShutdownTest(true);
+	}
+
+	@Test
+	public void testRequestResultBeforeSubmit() {
+		PerJobMiniClusterClient client = new PerJobMiniClusterClient(new Configuration(), mock(MiniCluster.class));
+		try {
+			client.requestJobResult(JobID.generate());
+			Assert.fail("This should have failed.");
+		} catch (IllegalStateException ignored) {
+			// That's what we want
+		}
+	}
+
+	@Test
+	public void testSubmitTwice() {
+		MiniCluster miniCluster = mock(MiniCluster.class);
+		PerJobMiniClusterClient client = new PerJobMiniClusterClient(new Configuration(), miniCluster);
+
+		JobGraph jobGraph = new JobGraph("test");
+		when(miniCluster.submitJob(jobGraph)).thenReturn(new CompletableFuture<>());
+		client.submitJob(jobGraph);
+		try {
+			client.submitJob(jobGraph);
+			Assert.fail("This should have failed.");
+		} catch (IllegalStateException ignored) {
+			// That's what we want
+		}
+	}
+
+	@Test
+	public void testRequestResultWithDifferentJobID() {
+		MiniCluster miniCluster = mock(MiniCluster.class);
+		PerJobMiniClusterClient client = new PerJobMiniClusterClient(new Configuration(), miniCluster);
+
+		JobGraph jobGraph = new JobGraph("test");
+		when(miniCluster.submitJob(jobGraph)).thenReturn(new CompletableFuture<>());
+		client.submitJob(jobGraph);
+		try {
+			client.requestJobResult(JobID.generate());
+			Assert.fail("This should have failed.");
+		} catch (IllegalStateException ignored) {
+			// That's what we want
+		}
+	}
+
+	private void runClusterShutdownTest(boolean withException) {
+		MiniCluster miniCluster = mock(MiniCluster.class);
+		CompletableFuture<JobSubmissionResult> submissionResult = new CompletableFuture<>();
+		CompletableFuture<JobResult> jobResult = new CompletableFuture<>();
+
+		JobGraph jobGraph = new JobGraph("test");
+		when(miniCluster.submitJob(jobGraph)).thenReturn(submissionResult);
+		when(miniCluster.requestJobResult(jobGraph.getJobID())).thenReturn(jobResult);
+
+		PerJobMiniClusterClient clusterClient = spy(new PerJobMiniClusterClient(new Configuration(), miniCluster));
+
+		CompletableFuture<JobID> jobIdFuture = clusterClient.submitJob(jobGraph);
+		// Trigger fetching the JobResult
+		jobIdFuture.complete(jobGraph.getJobID());
+
+		if (withException) {
+			jobResult.completeExceptionally(new RuntimeException("oups"));
+		} else {
+			jobResult.complete(mock(JobResult.class));
+		}
+
+		verify(clusterClient).shutDownCluster();
```

Comment:
We should try to reduce the usage of Mockito as it makes testing code harder to refactor. I think the problem here is that we don't have a proper `MiniCluster` interface which we could instrument. Maybe we should introduce this instead of using Mockito.

Reply:
I've drastically reduced the usage of mocks, there is just one test case left.